File: workbench/backend/schemas/_inspection.py
from __future__ import annotations

from typing import Literal

from pydantic import Field

from workbench.backend.schemas._base import (
    ApiResponseModel,
    BoundedIdentifier,
    ConfigOverrides,
    JsonObject,
    JsonValue,
)


class GraphConfigFieldResponse(ApiResponseModel):
    key: str
    value: JsonValue
    description: str | None = None


class GraphConfigResponse(ApiResponseModel):
    typeName: str
    fields: list[GraphConfigFieldResponse]


class GraphNodeResponse(ApiResponseModel):
    id: str
    label: str
    typeName: str
    description: str | None = None
    path: str
    graphRole: Literal["architecture", "internal", "runtime"]
    parameterCount: int
    parameterSizeBytes: int
    details: JsonObject
    config: GraphConfigResponse | None


class GraphEdgeResponse(ApiResponseModel):
    id: str
    source: str
    target: str


class InspectRequest(ApiResponseModel):
    modelType: BoundedIdentifier
    model: BoundedIdentifier
    preset: BoundedIdentifier
    overrides: ConfigOverrides = Field(default_factory=dict)
    experimentTask: BoundedIdentifier | None = None
    dataset: BoundedIdentifier | None = None
    logRunId: BoundedIdentifier | None = None


class InspectResponse(ApiResponseModel):
    modelType: str
    model: str
    preset: str
    parameterCount: int
    parameterSizeBytes: int
    nodes: list[GraphNodeResponse]
    edges: list[GraphEdgeResponse]

File: apps/workbench/web/src/lib/api/inspection.ts
import { z } from "zod";

import { requestJson } from "@/lib/api/client";
import { modelIdentitySegmentSchema } from "@/lib/api/model-identity-schema";
import {
  jsonObjectSchema,
  jsonValueSchema,
  type ConfigOverrides,
} from "@/lib/api/schemas";

export const graphConfigSchema = z.object({
  typeName: z.string(),
  fields: z.array(
    z.object({
      key: z.string(),
      value: jsonValueSchema,
      description: z.string().nullable().optional(),
    }),
  ),
});

export const graphNodeSchema = z.object({
  id: z.string(),
  label: z.string(),
  typeName: z.string(),
  description: z.string().nullable().optional(),
  path: z.string(),
  graphRole: z.enum(["architecture", "internal", "runtime"]),
  parameterCount: z.number(),
  parameterSizeBytes: z.number().default(0),
  details: jsonObjectSchema,
  config: graphConfigSchema.nullable(),
});

export const graphEdgeSchema = z.object({
  id: z.string(),
  source: z.string(),
  target: z.string(),
});

export const inspectResponseSchema = z.object({
  modelType: modelIdentitySegmentSchema,
  model: modelIdentitySegmentSchema,
  preset: z.string(),
  parameterCount: z.number(),
  parameterSizeBytes: z.number().default(0),
  nodes: z.array(graphNodeSchema),
  edges: z.array(graphEdgeSchema),
});

export type GraphConfig = z.infer<typeof graphConfigSchema>;
export type GraphNode = z.infer<typeof graphNodeSchema>;
export type GraphEdge = z.infer<typeof graphEdgeSchema>;
export type InspectResponse = z.infer<typeof inspectResponseSchema>;

export function inspectModel(input: {
  modelType: string;
  model: string;
  preset: string;
  overrides: ConfigOverrides;
  experimentTask?: string;
  dataset?: string;
  logRunId?: string;
}, options: { signal?: AbortSignal } = {}) {
  return requestJson("/inspect", inspectResponseSchema, {
    ...options,
    method: "POST",
    body: JSON.stringify(input),
  });
}
